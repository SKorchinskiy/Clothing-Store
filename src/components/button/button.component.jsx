import { DefaultButton, InvertedButton, GoogleButton } from "./button.styles";

function Button({ children, type, buttonType, onClick }) {
  const CustomButton = getStyledButtonComponent(buttonType);
  return (
    <CustomButton type={type} onClick={onClick}>
      {children}
    </CustomButton>
  );
}

function getStyledButtonComponent(buttonType = "default-btn") {
  return {
    "default-btn": DefaultButton,
    "invert-btn": InvertedButton,
    "google-btn": GoogleButton,
  }[buttonType];
}

export default Button;
