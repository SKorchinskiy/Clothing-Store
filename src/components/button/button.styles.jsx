import styled from "styled-components";

const BaseButton = styled.button`
  width: 100%;
  height: 50px;
  border-radius: 4px;
  border: 0;
  font-family: Josefin Slab, sans-serif;
  letter-spacing: 2px;
  font-size: large;
  cursor: pointer;

  &:hover {
    scale: 105%;
    user-select: none;
    transition-duration: 0.4s;
  }

  &:active {
    transition-duration: 0.1s;
    scale: 95%;
  }
`;

export const DefaultButton = styled(BaseButton)`
  background-color: rgba(0, 0, 0, 0.9);
  box-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
  color: white;

  &:hover {
    background-color: rgba(255, 255, 255, 0.9);
    color: black;
  }
`;

export const InvertedButton = styled(BaseButton)`
  background-color: rgba(255, 255, 255, 0.9);
  box-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
  color: black;

  &:hover {
    background-color: rgba(0, 0, 0, 0.9);
    color: white;
  }
`;

export const GoogleButton = styled(BaseButton)`
  background-color: rgb(66, 133, 244, 0.9);
  box-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
  color: white;
`;
