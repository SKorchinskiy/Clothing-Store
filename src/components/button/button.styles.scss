.btn-container {
  width: 100%;
  height: 50px;
  border-radius: 4px;
  border: 0;
  font-family: Josefin Slab, sans-serif;
  letter-spacing: 2px;
  font-size: large;
  cursor: pointer;

  &:hover {
    scale: 105%;
    user-select: none;
    transition-duration: 0.4s;
  }

  &:active {
    scale: 95%;
  }
}

.default-btn {
  background-color: rgba(0, 0, 0, 0.9);
  box-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
  color: white;

  &:hover {
    background-color: rgba(255, 255, 255, 0.9);
    color: black;
  }
}

.invert-btn {
  background-color: rgba(255, 255, 255, 0.9);
  box-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
  color: black;

  &:hover {
    background-color: rgba(0, 0, 0, 0.9);
    color: white;
  }
}

.google-btn {
  background-color: rgb(66, 133, 244, 0.9);
  box-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
  color: white;
}
