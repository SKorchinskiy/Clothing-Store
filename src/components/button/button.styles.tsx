import styled from "styled-components";

const BaseButton = styled.button`
  width: 100%;
  height: 50px;
  border-radius: 4px;
  border: 0;
  font-family: Josefin Slab, sans-serif;
  letter-spacing: 2px;
  font-size: large;
  margin-bottom: 10px;
  cursor: pointer;

  &:hover {
    opacity: 0.9;
    transition-duration: 0.3s;
  }

  @media screen and (max-width: 400px) {
    height: 40px;
    font-size: medium;
    padding: 5px;
  }
`;

export const DefaultButton = styled(BaseButton)`
  background-color: rgba(0, 0, 0, 0.9);
  box-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
  color: white;
`;

export const InvertedButton = styled(BaseButton)`
  background-color: rgba(255, 255, 255, 0.9);
  box-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
  color: black;

  @media screen and (min-width: 400px) {
    &:hover {
      transition: 0.5s;
      background-color: rgba(0, 0, 0, 0.9);
      color: white;
    }
  }
`;

export const GoogleButton = styled(BaseButton)`
  background-color: rgb(66, 133, 244, 0.9);
  box-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
  color: white;
`;
