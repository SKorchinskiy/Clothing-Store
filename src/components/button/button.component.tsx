import { DefaultButton, InvertedButton, GoogleButton } from "./button.styles";
import { MouseEventHandler } from "react";

export enum ButtonType {
  default = "default-btn",
  invert = "invert-btn",
  google = "google-btn",
}

type ButtonInput = {
  children: string;
  type: "button" | "submit";
  buttonType: ButtonType;
  onClick: MouseEventHandler<HTMLButtonElement>;
};

function Button({ children, type, buttonType, onClick }: ButtonInput) {
  const CustomButton = getStyledButtonComponent(buttonType);
  return (
    <CustomButton type={type} onClick={onClick}>
      {children}
    </CustomButton>
  );
}

function getStyledButtonComponent(buttonType: ButtonType = ButtonType.default) {
  return {
    "default-btn": DefaultButton,
    "invert-btn": InvertedButton,
    "google-btn": GoogleButton,
  }[buttonType];
}

export default Button;
