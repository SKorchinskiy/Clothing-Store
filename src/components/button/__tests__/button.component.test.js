import { screen, render, act } from "@testing-library/react";
import "@testing-library/jest-dom";

import Button from "../button.component";
import { DefaultButton, GoogleButton, InvertedButton } from "../button.styles";
import userEvent from "@testing-library/user-event";

describe("<Button/> component", () => {
  it("should be rendered in the document", () => {
    render(<Button />);

    const testButton = screen.getByRole("button");

    expect.assertions(1);
    expect(testButton).toBeInTheDocument();
  });

  it("should contain <DefaultButton />'s componentId", () => {
    const { asFragment } = render(<Button buttonType="default-btn" />);

    const testButton = screen.getByRole("button");

    expect.assertions(2);
    expect(testButton.className).toContain(
      DefaultButton.componentStyle.componentId
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should contain <InvertedButton />'s componentId", () => {
    const { asFragment } = render(<Button buttonType="invert-btn" />);

    const testButton = screen.getByRole("button");

    expect.assertions(2);
    expect(testButton.className).toContain(
      InvertedButton.componentStyle.componentId
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should contain <GoogleButton />'s componentId", () => {
    const { asFragment } = render(<Button buttonType="google-btn" />);

    const testButton = screen.getByRole("button");

    expect.assertions(2);
    expect(testButton.className).toContain(
      GoogleButton.componentStyle.componentId
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render child component", () => {
    const child = <span>Button's child</span>;
    render(<Button children={child} />);

    const textElement = screen.getByText("Button's child");

    expect.assertions(1);
    expect(textElement).toBeInTheDocument();
  });

  it("should call event handler when clicked", async () => {
    const mockedHandler = jest.fn();
    render(<Button onClick={mockedHandler} />);

    const testButton = screen.getByRole("button");
    await act(async () => await userEvent.click(testButton));

    expect.assertions(1);
    expect(mockedHandler).toBeCalled();
  });
});
