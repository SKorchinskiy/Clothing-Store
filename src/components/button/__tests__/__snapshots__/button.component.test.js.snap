// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> component should contain <DefaultButton />'s componentId 1`] = `
<DocumentFragment>
  <button
    class="sc-bcPKhP sc-grXZZQ cbtdGf fkLpdX"
  />
</DocumentFragment>
`;

exports[`<Button/> component should contain <GoogleButton />'s componentId 1`] = `
<DocumentFragment>
  <button
    class="sc-bcPKhP sc-hLclGa cbtdGf loojds"
  />
</DocumentFragment>
`;

exports[`<Button/> component should contain <InvertedButton />'s componentId 1`] = `
<DocumentFragment>
  <button
    class="sc-bcPKhP sc-dkjKgF cbtdGf kNOHCT"
  />
</DocumentFragment>
`;
